### ISSUE TITLE
[Bug]: useEditor always creates at least 2 editors on mount

### ISSUE THREAD

**Issue by schontz** [ueberdosis/tiptap#5432]
### Affected Packages

react

### Version(s)

develop

### Bug Description

The editor is [destroyed and re-created whenever deps changes](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L155-L162). But this `useEffect` will run once on mount to start, before `deps` has actually changed. If `deps` are supplied, then the editor created in `useState` will get destroyed _immediately_ in `useEffect` and get re-created.

It may be generally useful to have a hook that only runs when things change, e.g.:

```javascript
function useChangeEffect(effect, deps) {
  const isFirstTime = useRef(true);
  useEffect(() => {
    if (isFirstTime.current) isFirstTime.current = false;
    else return effect();
  }, deps);
}
```

Then you can just replace [this useEffect](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L123) with `useChangeEffect`.

On a related note, [applying options instead of re-creating an editor is useless code](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L147-L153), because the effect only runs when `deps` change, not when `options` change. That means updated `options` are never injected into the editor. Furthermore, [mostRecentOptions](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L80) is a misnomer, because it is not actually updated on each render, so it is effectively useless.

### Browser Used

Chrome

### Expected Behavior

Editor is only created once if deps do not change.

### Dependency Updates

- [X] Yes, I've updated all my dependencies.

**Comment by nperez0111** [comment]
Thanks for bringing it up, I was a bit too focused on another bug. I'll consider your some of your suggestions.

**Comment by schontz** [comment]
Looking at `useEditor` further, it doesn't even look like [deleteUnusedEditor](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L124-L137) will ever work, because `mostRecentEditor.current` will always point to the value of `editor`. 

1. Create `editor` via `useState`
2. `useEffect` will [destroy right away](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L124-L137) and re-create on mount (this bug)
3. If deps changes, `deleteUnusedEditor` will run again, but the editor is current 
4. On unmount, `deleteUnusedEditor` will run, but it will not destroy the editor because [editorInstance _is_ the most recent editor](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L132)

I was working on a PR for this, but there are multiple issues with the logic and I worry a little about pulling the thread.

I'm not clear under what conditions [useEffect cleanup will run _before_ the editor has been replaced](https://github.com/ueberdosis/tiptap/blob/develop/packages/react/src/useEditor.ts#L130-L131), since it will be run _after_ deps change (which necessitates a new editor).

The change that introduced this was supposedly for a strict mode issue, but because we create the editor via `useState`, it will always be run 2x in strict mode. There's no real workaround, so that means an orphaned editor will be created on every mount in strict mode. I know people want immediate rendering to work, and perhaps we can bring this up with React core team, as I have tried may workarounds for this type of scenario, but I have yet to find a practical solution. I suppose leaking memory in dev isn't the end of the world, but it's still frustrating.

That said, it is more react-y if it is not available on sync and instead runs all the commands via a useEffect. For example [this bug](https://github.com/ueberdosis/tiptap/issues/5001) wants to focus the editor right away, but in React the DOM isn't available until effects run, so the prescribed method should be something like:

```javascript
const editor = useEditor(...)

useEffect(() => {
  if (!editor) return;
  editor.commands.focus();
}, [editor]);

// or via onCreate
const editor = useEditor({
  onCreate: ({ editor }) => editor.commands.focus()
})
```

Although there are scenarios when the editor gets re-created, which would call that effect multiple times. This could be solved via an option, though a properly descriptive name is hard to come by:

```javascript
const editor = useEditor({
  onFirstTimeCreate: ({ editor }) => editor.commands.focus()
})
```

There could be more discussion around the API as a whole. For example, `useEditor` _could_ return a ref instead of a state, assuming it is available on mount:

```
const editorRef = useEditor(...)

useEffect(() => {
  // editorRef.current is set, much like DOM refs
  const editor = editorRef.current;
  if (!editor) return;
  editor.commands.focus();
}, [])
```

I know that doesn't play well with `deps`, as the editor ref may change over time. One might argue that the actual editor instance changing should not be something that you need to react to (assuming state is preserved between instances), and instead the editor should only be a ref that is reacted to via `useEditorState`. 

I put together a little [StackBlitz Sample](https://stackblitz.com/edit/vitejs-vite-ku94rh?file=src%2FApp.jsx) that shows some of these issues with a fake editor in strict and non strict modes. You can see that in both modes, the timeout destroy is not running on dismount.

**Comment by schontz** [comment]
Note that I am able to apply some of the patches locally and it stops creating new editors, though we'll have to refactor the "useState for new editor" if we really want things to be safe in React land.

**Comment by rfgamaral** [comment]
@nperez0111 Did this issue not get fixed with #5445? Is there a reason why the issue remains open?

**Comment by nperez0111** [comment]
Working on releasing it now actually 😅 

My CI has a bug that kept trying to release it as a minor that I was fighting. Will get it out as a patch

**Comment by nperez0111** [comment]
This is now released with v2.5.9!
